src/Development/IDE/Core/Completions.hs

```diff
@@ -0,0 +1,565 @@
+-- Mostly taken from "haskell-ide-engine"
+module Development.IDE.Core.Completions (
+  CachedCompletions
+, cacheDataProducer
+, WithSnippets(..)
+,getCompletions
+) where
+
+import Control.Applicative
+import Data.Aeson
+import Data.Aeson.Types
+import Data.Char (isSpace)
+import Data.Generics
+import Data.List as List hiding (stripPrefix)
+import qualified Data.Map  as Map
+import Data.Maybe (fromMaybe, mapMaybe)
+import qualified Data.Text as T
+import qualified Text.Fuzzy as Fuzzy
+
+import GHC
+import Module
+import HscTypes
+import Name
+import RdrName
+import TcRnTypes
+import Type
+import Var
+import Packages
+import DynFlags
+import ConLike
+import DataCon
+import SrcLoc as GHC
+
+import Language.Haskell.LSP.Types
+import Language.Haskell.LSP.Types.Capabilities
+import qualified Language.Haskell.LSP.VFS as VFS
+import Development.IDE.Core.CompletionsTypes
+import Development.IDE.Spans.Documentation
+
+-- From haskell-ide-engine/src/Haskell/Ide/Engine/Support/HieExtras.hs
+
+data NameDetails
+  = NameDetails Module OccName
+  deriving (Eq)
+
+nsJSON :: NameSpace -> Value
+nsJSON ns
+  | isVarNameSpace ns = String "v"
+  | isDataConNameSpace ns = String "c"
+  | isTcClsNameSpace ns  = String "t"
+  | isTvNameSpace ns = String "z"
+  | otherwise = error "namespace not recognized"
+
+parseNs :: Value -> Parser NameSpace
+parseNs (String "v") = pure Name.varName
+parseNs (String "c") = pure dataName
+parseNs (String "t") = pure tcClsName
+parseNs (String "z") = pure tvName
+parseNs _ = mempty
+
+instance FromJSON NameDetails where
```

Comment:
All this JSON stuff is not needed here because it is only used for the `resolveCompletion` LSP request, where the types of non local names and documentation is looked up. It doesn't seem like you added support for the resolve completion request.

Reply:
Nope, I didn't implement the resolver completion request, since all information is already sent in the original response. Is this wrong?

Reply:
I don't think you are going to get types for things not defined in the active module, since they are looked up in the local type env.

Reply:
If you don't plan to implement the resolve completion request, you can remove everything to do with JSON - which is everything from line 42 to 77.